File: 064_Remove Duplicates from Sorted List II.cpp
/*
Remove Duplicates from Sorted List II

Delete all nodes that have duplicate numbers from a sorted linked list.
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode { int val; ListNode* next; ListNode(int x): val(x), next(nullptr) {} };

ListNode* deleteDuplicates(ListNode* head) {
    ListNode dummy(0); dummy.next = head;
    ListNode* prev = &dummy;
    while (prev->next) {
        ListNode* curr = prev->next;
        bool dup = false;
        while (curr->next && curr->val == curr->next->val) {
            dup = true;
            ListNode* tmp = curr->next; curr->next = tmp->next; delete tmp;
        }
        if (dup) { // remove curr as well
            ListNode* tmp = curr;
            prev->next = curr->next;
            delete tmp;
        } else prev = prev->next;
    }
    return dummy.next;
}

void printList(ListNode* h) { while (h) { cout << h->val; if (h->next) cout << ' '; h = h->next; } cout << endl; }

int main() {
    ListNode* a = new ListNode(1); a->next = new ListNode(2); a->next->next = new ListNode(3); a->next->next->next = new ListNode(3); a->next->next->next->next = new ListNode(4);
    auto r = deleteDuplicates(a); // expected 1 2 4
    cout << "DelDup 1: "; printList(r);

    ListNode* b = new ListNode(1); b->next = new ListNode(1); b->next->next = new ListNode(1); b->next->next->next = new ListNode(2); b->next->next->next->next = new ListNode(3);
    cout << "DelDup 2: "; printList(deleteDuplicates(b)); // 2 3
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 064_Remove Duplicates from Sorted List II.py
"""
Remove Duplicates from Sorted List II

Delete all nodes that have duplicate numbers from a sorted linked list.
"""

class ListNode:
    def __init__(self,x):
        self.val = x
        self.next = None

def delete_duplicates(head: ListNode) -> ListNode:
    dummy = ListNode(0); dummy.next = head
    prev = dummy
    while prev.next:
        curr = prev.next
        dup = False
        while curr.next and curr.val == curr.next.val:
            dup = True
            curr.next = curr.next.next
        if dup:
            prev.next = curr.next
        else:
            prev = prev.next
    return dummy.next

def print_list(h: ListNode):
    print(' '.join(str(x) for x in iter_list(h)))

def iter_list(h: ListNode):
    while h:
        yield h.val; h = h.next

if __name__ == '__main__':
    a = ListNode(1); a.next = ListNode(2); a.next.next = ListNode(3); a.next.next.next = ListNode(3); a.next.next.next.next = ListNode(4)
    print('DelDup 1:', end=' '); print_list(delete_duplicates(a))
    b = ListNode(1); b.next = ListNode(1); b.next.next = ListNode(1); b.next.next.next = ListNode(2); b.next.next.next.next = ListNode(3)
    print('DelDup 2:', end=' '); print_list(delete_duplicates(b))

# Time Complexity: O(n)
# Space Complexity: O(1)
